programs/axelar-solana-gas-service-v2/src/instructions/add_spl_gas.rs

```diff
@@ -0,0 +1,96 @@
+use crate::state::Config;
+use anchor_lang::prelude::*;
+use anchor_spl::token_interface::{self, Mint, TokenAccount, TokenInterface, TransferChecked};
+use axelar_solana_gas_service_events::events::SplGasAddedEvent;
+
+/// Add more gas (SPL tokens) to an existing contract call.
+///
+/// Accounts expected:
+/// 0. `[signer, writable]` The account (`sender`) paying the gas fee in SPL tokens.
+/// 1. `[writable]` The sender's associated token account for the mint.
+/// 2. `[writable]` The `config_pda` account.
+/// 3. `[writable]` The config PDA's associated token account for the mint.
+/// 4. `[]` The mint account for the SPL token.
+/// 5. `[]` The SPL token program.
+/// 6+. `[signer, writable]` Optional additional accounts required by the SPL token program for the transfer.
+#[event_cpi]
+#[derive(Accounts)]
+pub struct AddSplGas<'info> {
+    #[account(mut)]
+    pub sender: Signer<'info>,
+
+    #[account(
+        mut,
+        associated_token::mint = mint,
+        associated_token::authority = sender,
+    )]
+    pub sender_ata: InterfaceAccount<'info, TokenAccount>,
+
+    #[account(
+        mut,
+        seeds = [
+            Config::SEED_PREFIX,
+        ],
+        bump = config_pda.load()?.bump,
+    )]
+    pub config_pda: AccountLoader<'info, Config>,
+
+    #[account(
+        mut,
+        associated_token::mint = mint,
+        associated_token::authority = config_pda,
+    )]
+    pub config_pda_ata: InterfaceAccount<'info, TokenAccount>,
+
+    pub mint: InterfaceAccount<'info, Mint>,
+
+    pub token_program: Interface<'info, TokenInterface>,
+}
+
+pub fn add_spl_gas<'info>(
+    // explicitly specify all lifetimes to fix the `remaining_accounts` issue
+    // see more: https://solana.stackexchange.com/questions/13275/cpicontext-with-remaining-accounts-is-not-working-because-of-lifetimes
+    ctx: Context<'_, '_, '_, 'info, AddSplGas<'info>>,
+    tx_hash: [u8; 64],
+    log_index: u64,
+    gas_fee_amount: u64,
+    decimals: u8,
+    refund_address: Pubkey,
+) -> Result<()> {
+    if gas_fee_amount == 0 {
+        msg!("Gas fee amount cannot be zero");
+        return Err(ProgramError::InvalidInstructionData.into());
+    }
+
+    // TODO(v2) check if this check is necessary
+    // or move to account constraint
+    if decimals != ctx.accounts.mint.decimals {
+        msg!("Decimals do not match the mint's decimals");
+        return Err(ProgramError::InvalidInstructionData.into());
+    }
```

Comment:
```suggestion
```
IIRC, the difference between the `Transfer` and `TransferChecked` instructions is that the `TransferChecked` checks if the decimals given match the mint decimals ([source](https://github.com/solana-program/token-2022/blob/main/program/src/processor.rs#L348)), therefore this would be redundant.

Reply:
@frenzox thanks!